### PR TITLE
bpf: allow to enable host_routing and endpoint routes simultaneously

### LINKED ISSUE #14240: cilium: endpoint routes + veth fast redirect support
Both settings should work with fast redirect for 5.10+.

### PR DESCRIPTION
**Pull request by aspsk**
What the title says, see the commit message for details

Fixes: #14240

```release-note
Enable endpoint routes + veth fast redirect support
```

### CODE REVIEW

**Comment by aspsk** [comment]
/test-1.25-net-next

**Comment by aspsk** [comment]
/test-1.25-net-next

File: bpf/bpf_lxc.c [hunk 1]
```diff
@@ -973,19 +979,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		 */
 		ep = lookup_ip4_endpoint(ip4);
 		if (ep) {
-#ifdef ENABLE_ROUTING
 			if (ep->flags & ENDPOINT_F_HOST) {
+				if (is_defined(ENABLE_ROUTING)) {
```
Comment (julianwiedmann) [inline bpf/bpf_lxc.c:1049]: Something that triggers me whenever I see this code - is the check for `ENDPOINT_F_HOST` actually needed? After we jump to `to_host`, we end up checking `*dst_id == HOST_ID` instead. Are the two conditions completely interchangeable?

If so we could also check for `HOST_ID` here, avoid the endpoint-lookup, and make the overall logic a bit simpler.

**Comment by aspsk** [comment]
/test-1.25-4.19

File: bpf/bpf_lxc.c (hunk 1 above)
Comment (aspsk) [reply bpf/bpf_lxc.c:1049]: Good point. The `dst_id` is taken from the `ipcache` map, so could it happen that the entry doesn't exist for an existing endpoint?

I wouldn't try to patch this in the context of this PR in any case (also given that we're close to the freeze)

File: pkg/datapath/linux/config/config.go [hunk 2]
```diff
@@ -863,7 +863,7 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 		fmt.Fprintf(fw, "#define USE_BPF_PROG_FOR_INGRESS_POLICY 1\n")
 	}
 
-	if option.Config.ForceLocalPolicyEvalAtSource {
+	if option.Config.ForceLocalPolicyEvalAtSource || (option.Config.EnableEndpointRoutes && !option.Config.EnableHostLegacyRouting) {
```
Comment (pchaigno) [inline pkg/datapath/linux/config/config.go]: IIUC, this is okay and won't end up causing https://github.com/cilium/cilium/issues/14657 because with BPF Host Routing, we'll skip the second policy enforcement. Is that correct?

I think I would prefer to have that bit of logic in the datapath (so using the two corresponding macros) with a big comment to explain why we're doing that and why it's fine.

File: bpf/bpf_lxc.c [hunk 3]
```diff
@@ -451,19 +452,19 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 		 */
 		ep = lookup_ip6_endpoint(ip6);
 		if (ep) {
-#ifdef ENABLE_ROUTING
 			if (ep->flags & ENDPOINT_F_HOST) {
+				if (is_defined(ENABLE_ROUTING)) {
```
Comment (pchaigno) [inline bpf/bpf_lxc.c:573]: Why is this not an #ifdef anymore?

File: bpf/bpf_lxc.c [hunk 4]
```diff
@@ -972,19 +978,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
 #ifdef HOST_IFINDEX
-				goto to_host;
-#else
-				return DROP_HOST_UNREACHABLE;
+					goto to_host;
 #endif
+					return DROP_HOST_UNREACHABLE;
+				}
+				goto pass_to_stack;
```
Comment (pchaigno) [inline bpf/bpf_lxc.c:1055]: I'm trying to understand this change in behavior, but I'm not sure I understand how it even used to work. AFAICS, when endpoint routes were enabled, we used to deliver packets going to the host via `ipv4_local_delivery` but that will fail as `ep->lxc_id` is 0 for the host endpoint :confused:

File: bpf/lib/fib.h [hunk 5]
```diff
@@ -10,6 +10,12 @@
 #include "common.h"
 #include "l3.h"
 
+/*
+ * Translation into human language: do a fib lookup if there is more than one
+ * output device or per-endpoint routes are enabled (which => !ENABLE_ROUTING)
+ */
+#define DO_FIB_LOOKUP (!is_defined(ENABLE_SKIP_FIB) || !is_defined(ENABLE_ROUTING))
```
Comment (pchaigno) [inline bpf/lib/fib.h]: I'd make it `ENABLE_FIB_LOOKUP` if possible, to be consistent with other macros.

File: bpf/lib/fib.h [hunk 6]
```diff
@@ -10,6 +10,12 @@
 #include "common.h"
 #include "l3.h"
 
+/*
+ * Translation into human language: do a fib lookup if there is more than one
+ * output device or per-endpoint routes are enabled (which => !ENABLE_ROUTING)
```
Comment (pchaigno) [inline bpf/lib/fib.h]: Can we get the *why* instead of the *what*? :grin: 
I'm currently missing why your patch needs to change this.

**Review by pchaigno — COMMENTED** [review]
Couple comments below. All in all, this is tricky logic so I think it warrants a few more explanations :slightly_smiling_face:

File: bpf/bpf_lxc.c (hunk 4 above)
Comment (borkmann) [reply bpf/bpf_lxc.c:1055]: > when endpoint routes were enabled, we used to deliver packets going to the host via ipv4_local_delivery

Don't we pass everything up the stack with per endpoint routes? The ipv{4,6}_local_delivery() was used for !per-endpoint routes case to deliver into Pods.

Comment (pchaigno) [reply bpf/bpf_lxc.c:1055]: Well, I guess that's part of the weird thing here. If you look at the code as it is in master today, you can see that the call to `ipv4_local_delivery` (3 lines below) was outside the `ENABLE_ROUTING` guard. So we were (and still are) calling `ipv4_local_delivery` regardless of the endpoint routes setting it seems :confused:

Comment (aspsk) [reply bpf/bpf_lxc.c:1055]: In master this was actually protected by the `is_defined(ENABLE_ROUTING) || hairpin`, so in case of endpoint routes this piece of code is skipped

File: bpf/bpf_lxc.c (hunk 3 above)
Comment (aspsk) [reply bpf/bpf_lxc.c:573]: I've copy-pasted this from the original version of this patch, where Daniel used this to make code more readable

File: pkg/datapath/linux/config/config.go (hunk 2 above)
Comment (aspsk) [reply pkg/datapath/linux/config/config.go]: > IIUC, this is okay and won't end up causing #14657 because with BPF Host Routing, we'll skip the second policy enforcement. Is that correct?

Yes, we jump right into the pod namespace.

> I think I would prefer to have that bit of logic in the datapath (so using the two corresponding macros) with a big comment to explain why we're doing that and why it's fine.

```
/*
 * When the host routing is enabled we need to check policies at source.
 * Otherwise, as the skb will be delivered directly to pod's namespace,
 * the ingress policy (the cil_to_container BPF program) will be bypassed.
 */
#if defined(ENABLE_ENDPOINT_ROUTES) && defined(ENABLE_HOST_ROUTING)
#  ifndef FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
#  define FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
#  endif
#endif
```
Does this work for you @pchaigno ?

Comment (pchaigno) [reply pkg/datapath/linux/config/config.go]: Yep. Thanks!

File: bpf/bpf_lxc.c (hunk 3 above)
Comment (pchaigno) [reply bpf/bpf_lxc.c:573]: cc @borkmann

File: bpf/bpf_lxc.c [hunk 7]
```diff
@@ -956,11 +957,16 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
-	 * disabled. Otherwise, the packet will be dropped by the kernel if
-	 * it is going to be routed via an interface it came from after it has
-	 * been passed to the stack.
+	 * disabled (e.g. the case in per-endpoint routes). Otherwise, the
+	 * packet will be dropped by the kernel if it is going to be routed via
+	 * an interface it came from after it has been passed to the stack.
+	 *
+	 * For the case where ENABLE_ROUTING is disabled, but the fast redirect
+	 * enabled we do lookup the local endpoint here to check whether we
+	 * must pass the packet up the stack for the host itself.
```
Comment (pchaigno) [inline bpf/bpf_lxc.c]: That's not the only reason to lookup the local endpoint, no? We also want to run through `ipv4_local_delivery` to enforce ingress policies for that endpoint.

Comment (aspsk) [reply bpf/bpf_lxc.c]: Yes, thanks, I will update this comment

File: bpf/bpf_lxc.c (hunk 4 above)
Comment (pchaigno) [reply bpf/bpf_lxc.c:1055]: Oh, right :facepalm: I was only looking at the ifdefs.

I think I get the change now. The old check on `ENABLE_ROUTING` (line 975) was useless then, right?

Comment (aspsk) [reply bpf/bpf_lxc.c:1055]: The old one was to optimize code for the hairpin case, I guess

File: bpf/lib/fib.h (hunk 6 above)
Comment (aspsk) [reply bpf/lib/fib.h]: @pchaigno I've replaced this with a proper macro

**Comment by aspsk** [comment]
/test-1.26-net-next

File: bpf/lib/fib.h (hunk 6 above)
Comment (julianwiedmann) [reply bpf/lib/fib.h]: +1 to what @pchaigno said - this needs an explanation *why* per-EP routes now always requires a FIB lookup.

File: bpf/bpf_lxc.c [hunk 8]
```diff
@@ -973,19 +981,19 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
+				if (is_defined(ENABLE_ROUTING)) {
 #ifdef HOST_IFINDEX
-				goto to_host;
-#else
-				return DROP_HOST_UNREACHABLE;
+					goto to_host;
 #endif
+					return DROP_HOST_UNREACHABLE;
+				}
+				goto pass_to_stack;
```
Comment (julianwiedmann) [inline bpf/bpf_lxc.c:1055]: For `ENDPOINT_F_HOST` and per-EP routes, this jump to `pass_to_stack` means we're bypassing quite a bit of code. Can you explain why it's ok to bypass the [section right below](https://github.com/cilium/cilium/blob/fcf2b7ad1c47cc7309e752eb087ecfbda7ff8a6e/bpf/bpf_lxc.c#L992-L1001) (Host-Ingress Policy Enforcement with HostFW + per-EP routes)? If it's no longer relevant, can it be removed?

**Comment by joestringer** [comment]
@cilium/sig-datapath / @aspsk / @borkmann do we *really* need this for 1.13? I'm excited for this, but given the 1.13 release timeline we're running up to the wire on this one. We nominated this as a release-blocker just prior to freezing any further features, with the plan that if it continues to make active progress then we can merge it. I'm not seeing recent activity, there's a few unresolved comments, and it's not clear to me how this is being tested. At this point this is feeling like a bit of a risky change to make so late in the cycle.

**Comment by joestringer** [comment]
OK thanks, no worries. I will drop the 1.13 labels for now given your response above. We can always have a subsequent discussion about releases to include this in, but it should be viable at least in the 1.14 timeframe.

**Comment by nirousseau** [comment]
Hello,

Thank you for this contribution :-) (and https://github.com/JrCs for the pointer).

How can I test this "experimental" release on an OKD (Openshift) cluster ? Do i have to manually set the images of the Cilium components ? Can you confirm that I must be in "direct routing" mode (tunnel disabled along with legacy routing flag) ? Is OKD 4.11 a good candidate for testing ?

Regards,
Nikita

**Comment by aspsk** [comment]
/test-1.26-net-next

File: bpf/lib/l3.h [hunk 9]
```diff
@@ -14,6 +14,17 @@
...
+/*
+ * When the host routing is enabled we need to check policies at source.
+ * Otherwise, as the skb will be delivered directly to pod's namespace,
+ * the ingress policy (the cil_to_container BPF program) will be bypassed.
+ */
+#if defined(ENABLE_ENDPOINT_ROUTES) && defined(ENABLE_HOST_ROUTING)
+#  ifndef FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
+#  define FORCE_LOCAL_POLICY_EVAL_AT_SOURCE
+#  endif
+#endif
```
Comment (borkmann) [inline bpf/lib/l3.h:26]: I would move this section out of the C code and into the agent. The agent could enable option.Config.ForceLocalPolicyEvalAtSource when it knows that option.Config.EnableEndpointRoutes is true and option.Config.EnableHostLegacyRouting is false. I'd also move the above comment to that code location where you set this in the agent. Less ifdef'ery then for the C code.

File: daemon/cmd/kube_proxy_replacement.go [hunk 10]
```diff
@@ -449,8 +449,6 @@ func finishKubeProxyReplacementInit() error {
 		case option.Config.IptablesMasqueradingEnabled():
 			msg = fmt.Sprintf("BPF host routing requires %s.", option.EnableBPFMasquerade)
 		// All cases below still need to be implemented ...
```
Comment (borkmann) [inline daemon/cmd/kube_proxy_replacement.go:459]: nit: I think this comment above could then also be removed. 🎉

File: bpf/bpf_lxc.c [hunk 11]
```diff
@@ -527,8 +527,20 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
+#if defined(ENABLE_HOST_FIREWALL) && !defined(ENABLE_ROUTING)
+	/* If the destination is the local host and per-endpoint routes are
+	 * enabled, jump to the bpf_host program to enforce ingress host policies.
+	 */
+	if (*dst_id == HOST_ID) {
+		ctx_store_meta(ctx, CB_FROM_HOST, 0);
+		tail_call_static(ctx, &POLICY_CALL_MAP, HOST_EP_ID);
+		return DROP_MISSED_TAIL_CALL;
+	}
+#endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
```
Comment (borkmann) [inline bpf/bpf_lxc.c:556]: nit: Could we split out this move for HF above from the main commit for v4/v6 bits? Would be nice to move this out so the main commit shrinks a bit more, and for the HF we just do the move as a prep commit with no change in functionality.

**Review by borkmann — APPROVED** [review]
LGTM, just few nits below that would be nice to address before merging.

**Comment by aspsk** [comment]
/test-1.25-4.19

**Review by borkmann — APPROVED** [review]

File: bpf/bpf_lxc.c [hunk 12]
```diff
@@ -563,7 +563,8 @@ static __always_inline int handle_ipv6_from_lxc(struct __ctx_buff *ctx, __u32 *d
 #endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
```
Comment (pchaigno) [inline bpf/bpf_lxc.c:556]: From commit description:
> In order to enforce Host Firewall (CCNP), we move a corresponding code section
in bpf_lxc.c one block above.

I believe this needs to be updated because it was moved to its own commit (:pray:)

File: pkg/datapath/linux/config/config.go [hunk 13]
```diff
@@ -874,7 +874,9 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 		fmt.Fprintf(fw, "#define USE_BPF_PROG_FOR_INGRESS_POLICY 1\n")
 	}
 
-	if option.Config.ForceLocalPolicyEvalAtSource {
+	// When both the host routing and endpoint routes are enabled we need
+	// to check policies at source.
+	if option.Config.ForceLocalPolicyEvalAtSource || (option.Config.EnableEndpointRoutes && !option.Config.EnableHostLegacyRouting) {
 		fmt.Fprintf(fw, "#define FORCE_LOCAL_POLICY_EVAL_AT_SOURCE 1\n")
```
Comment (pchaigno) [inline pkg/datapath/linux/config/config.go]: Setting this in the global datapath configuration will force policy evaluation at the source for lxc -> lxc, but also for hostns -> lxc, eth0 -> lxc, and overlay -> lxc because all of those paths call `ipv{4,6}_local_delivery`. Is that really what we want?

AFAICS, that will cause us again to enforce policies twice for all of those paths, when endpoint routes and BPF Host Routing are enabled.

**Review by pchaigno — CHANGES_REQUESTED** [review]
I might be missing something, but AFAICS, this will reintroduce bug https://github.com/cilium/cilium/issues/14657. Please see comment below.

There's also one comment from Julian that wasn't resolved/answered above.

File: bpf/bpf_lxc.c (hunk 8 above)
Comment (aspsk) [reply bpf/bpf_lxc.c:1055]: This was addressed by moving host firewall check above

File: pkg/datapath/linux/config/config.go (hunk 13 above)
Comment (aspsk) [reply pkg/datapath/linux/config/config.go]: Thanks for the comments @pchaigno!

The idea is that `!option.Config.EnableHostLegacyRouting` means that after checking policy at source we will forward the packet directly to the pod using `redirect_peer`, thus bypassing the outer part of the veth. The only case when we trigger the policy program at the outer veth is when traffic gets there directly from the hostns (as endpoint routes are set). Does this make sense for you?

Comment (pchaigno) [reply pkg/datapath/linux/config/config.go]: Ok, I think I see what I was missing. The packet will then follow the path: bpf_lxc:handle_policy() -> tail_ipv4_ct_ingress_policy_only() -> tail_ipv4_policy() -> ipv4_policy() -> common.h:redirect_ep() -> ctx_redirect_peer(). And for hostns -> lxc, it won't skip the destination's ingress BPF program because of the `needs_backlog` variable, is that right?

Overall, it feels super fragile. The path is complex, with many conditions that could send us into a double enforcement or policy bypass if they ever change the wrong way.

What kind of test coverage do we have for this? Will we notice if `FORCE_LOCAL_POLICY_EVAL_AT_SOURCE` disappears tomorrow (consequence: policy bypass), or if `from_host` is not correctly propagated into `needs_backlog` anymore (consequence: policy bypass), or if [this condition](https://github.com/cilium/cilium/blob/4310b5e5c4a15049d963cbad8791226567743e9c/bpf/bpf_lxc.c#L1881) is extended (consequence: double enforcement)?

**Comment by aspsk** [comment]
/test-1.24-5.4

**Comment by aspsk** [comment]
/test-1.26-net-next

**Comment by julianwiedmann** [comment]
Note that this will need coordination with https://github.com/cilium/cilium/pull/24727 re `FORCE_LOCAL_POLICY_EVAL_AT_SOURCE`.

File: bpf/bpf_lxc.c [hunk 14]
```diff
@@ -1003,12 +1008,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
+	}
+#endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
+
 	/* Allow a hairpin packet to be redirected even if ENABLE_ROUTING is
-	 * disabled. Otherwise, the packet will be dropped by the kernel if
-	 * it is going to be routed via an interface it came from after it has
-	 * been passed to the stack.
+	 * disabled (e.g. the case in per-endpoint routes). Otherwise, the
+	 * packet will be dropped by the kernel if it is going to be routed via
+	 * an interface it came from after it has been passed to the stack.
```
Comment (zacharysarah) [inline bpf/bpf_lxc.c]: Avoid Latin abbreviations.
Tighten up verb tenses for clarity. (Please review for technical accuracy.)

```suggestion
	 * disabled (for example, with per-endpoint routes). Otherwise, the
	 * packet will be dropped by the kernel if the packet will be routed through
	 * the interface it came from after the packet has been passed to the stack.
```

File: bpf/bpf_lxc.c [hunk 15]
```diff
@@ -1003,12 +1008,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
...
+	 *
+	 * For the case where ENABLE_ROUTING is disabled, but the fast redirect
+	 * enabled we do lookup the local endpoint here to check whether we
+	 * must pass the packet up the stack for the host itself. We also want
+	 * to run through ipv4_local_delivery to enforce ingress policies for
+	 * that endpoint.
```
Comment (zacharysarah) [inline bpf/bpf_lxc.c]: Who is "we" here in both instances? It's unclear which component(s) are acting.

```suggestion
	 * If ENABLE_ROUTING is disabled but fast redirect is
	 * enabled, we do lookup the local endpoint here to check whether we
	 * must pass the packet up the stack for the host itself. We also want
	 * to run through `ipv4_local_delivery` to enforce ingress policies for
	 * that endpoint.
```

File: bpf/lib/l3.h [hunk 16]
```diff
@@ -14,6 +14,17 @@
 #include "icmp6.h"
 #include "csum.h"
 
+/*
+ * When the host routing is enabled we need to check policies at source.
+ * Otherwise, as the skb will be delivered directly to pod's namespace,
+ * the ingress policy (the cil_to_container BPF program) will be bypassed.
```
Comment (zacharysarah) [inline bpf/lib/l3.h]: Who is "we" here? Which component(s) are checking policies at source?
Avoid future tense.
```suggestion
 * When the host routing is enabled we need to check policies at source.
 * Otherwise, the skb is delivered directly to pod's namespace and
 * the ingress policy (the cil_to_container BPF program) is bypassed.
```

File: pkg/datapath/linux/config/config.go [hunk 17]
```diff
@@ -874,7 +874,9 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 		fmt.Fprintf(fw, "#define USE_BPF_PROG_FOR_INGRESS_POLICY 1\n")
 	}
 
-	if option.Config.ForceLocalPolicyEvalAtSource {
+	// When both the host routing and endpoint routes are enabled we need
+	// to check policies at source.
```
Comment (zacharysarah) [inline pkg/datapath/linux/config/config.go]: Who is "we"?

**Review by zacharysarah — CHANGES_REQUESTED** [review]
@aspsk Good start! I have some questions about which component(s) are acting in specific instances.

**Comment by aspsk** [comment]
/test-1.26-net-next

**Comment by aspsk** [comment]
The `ConformanceEKS` triggers https://github.com/cilium/cilium/issues/24774

**Comment by aspsk** [comment]
The net-next failure consists of https://github.com/cilium/cilium/issues/24643 and https://github.com/cilium/cilium/issues/24687

File: pkg/datapath/linux/config/config.go (hunk 13 above)
Comment (aspsk) [reply pkg/datapath/linux/config/config.go]: Hi @pchaigno! We will notice if policy is not applied correctly via the connectivity test. However, we currently don't automatically test dual policy enforcement.

**Comment by aspsk** [comment]
I've only enabling this change patch if IPv6 is off. This is, in particular, due problems with IPv6 neighbour discovery (see, e.g., the https://github.com/cilium/cilium/issues/23910 issue).

File: pkg/datapath/linux/config/config.go (hunk 13 above)
Comment (pchaigno) [reply pkg/datapath/linux/config/config.go]: That will only be the case if this configuration is covered in some test somewhere. Is it? Do the Datapath Conformance tests enable BPF Host Routing?

**Review by pchaigno — CHANGES_REQUESTED** [review]
Blocking on above testing discussion. This is fairly fragile, so we really need to make sure this is covered somewhere in CI. AFAICS, that is not the case today.

It may be as simple as extending the Conformance Datapath workflow.

File: pkg/datapath/linux/config/config.go (hunk 13 above)
Comment (aspsk) [reply pkg/datapath/linux/config/config.go]: The host routing is enabled by default unless it is disabled on kernels which don't support it (= `EnableHostLegacyRouting` is disabled by default.) The net-next, at least, runs with the host routing enabled.

Comment (pchaigno) [reply pkg/datapath/linux/config/config.go]: Discussed offline. Today the net-next tests automatically disable BPF Host Routing because it is not supported. Once this PR is merged, this "auto-disablement" won't happen anymore and the tests will start testing BPF Host Routing with endpoint routes without any changes to the tests themselves being needed.

LGTM.

**Review by pchaigno — APPROVED** [review]
Above discussion is resolved. LGTM!

**Comment by aspsk** [comment]
The EKS is still the same https://github.com/cilium/cilium/issues/24774

**Comment by aspsk** [comment]
The **62** failures in net-next is
```
/home/jenkins/workspace/Cilium-PR-K8s-1.26-kernel-net-next/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:453
failed due to BeforeAll failure
/home/jenkins/workspace/Cilium-PR-K8s-1.26-kernel-net-next/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:652
```
and
```
/home/jenkins/workspace/Cilium-PR-K8s-1.26-kernel-net-next/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:453
failed to ensure the namespace kube-system exists: signal: killed
/home/jenkins/workspace/Cilium-PR-K8s-1.26-kernel-net-next/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:652
```
So this is https://github.com/cilium/cilium/issues/24643 

One is https://github.com/cilium/cilium/issues/24687

The last run had only 3 flakes in net-next. Since then code didn't change much (removed the setting of FORCE_POLICY_AT_SOURCE, but we set it in `l3`, so nothing really changed).

So unless there are concerns I think this is ready to be merged.
